lib/raty.js

```diff
@@ -0,0 +1,784 @@
+  cancelHint: 'Cancel this rating!',
+  cancelOff: 'cancel-off.png',
+  cancelOn: 'cancel-on.png',
+  cancelPlace: 'left',
+  click: undefined,
+  half: false,
+  halfShow: true,
+  hints: ['bad', 'poor', 'regular', 'good', 'gorgeous'],
+  iconRange: undefined,
+  iconRangeSame: false,
+  mouseout: undefined,
+  mouseover: undefined,
+  noRatedMsg: 'Not rated yet!',
+  number: 5,
+  numberMax: 20,
+  path: undefined,
+  precision: false,
+  readOnly: false,
+  round: { down: 0.25, full: 0.6, up: 0.76 },
+  score: undefined,
+  scoreName: 'score',
+  single: false,
+  space: true,
+  starHalf: 'star-half.png',
+  starOff: 'star-off.png',
+  starOn: 'star-on.png',
+  starType: 'img',
+  target: undefined,
+  targetFormat: '{score}',
+  targetKeep: false,
+  targetScore: undefined,
+  targetText: '',
+  targetType: 'hint',
+});
+
+
+class Raty {
+  /**
+   *
+   * @param {object} options
+   * @param {string} element
+   *
+   */
+  constructor(element, options) {
+    //overwrite basic configuration
+    if (options) {
+      if (typeof options === 'object') {
+        this.opt = {
+          ...defaultoptions,
+          ...options,
+        };
+      } else {
+        throw new Error('Illegal parameter');
+      }
+    } else {
+      this.opt = defaultoptions;
+    }
+    this.self = document.querySelector(element);
+    this.element = element;
+    this.init();
+    //string to boolean
+    String.prototype.toBoolean = function () {
+      let str = this.toLowerCase().trim();
+      console.log(str);
+      try{
+        return JSON.parse(str);
+      }catch(err){
+        return;
+      }
+    };
+  }
+  cancel(click) {
+    if (typeof this.self.dataset.readOnly==="string" && !this.self.dataset.readOnly.toBoolean()) {
+      this[click ? 'click' : 'score'].call(this, null);
+      this.scoreField.removeAttribute('value');
+    }
+  }
+
+  // TODO: model spec
+  click(score) {
+    if (typeof this.self.dataset.readOnly === "string" && !this.self.dataset.readOnly.toBoolean()) {
+      score = this._adjustedScore(score);
+
+      this._apply(score);
+
+      if (this.opt.click) {
+        this.opt.click.call(this.self, score, new Event('click'));
+      }
+
+      this._target(score);
+    }
+  }
+
+  // TODO: model spec
+  move(score) {
+    var integer = parseInt(score, 10);
+    var decimal = this._getDecimal(score, 1);
+
+    if (integer >= this.opt.number) {
+      integer = this.opt.number - 1;
+      decimal = 10;
+    }
+
+    var width = this._getWidth();
+    var steps = width / 10;
+    var star = this.stars[integer];
+    var percent = star.offsetLeft + steps * decimal;
+    var evt = new Event('mousemove');
+    evt.pageX = percent;
+    this.isMove = true;
+    star.dispatchEvent(evt);
+    this.isMove = false;
+  }
+
+  // TODO: model spec
+  readOnly(readonly) {
+    if (typeof this.self.dataset.readOnly === "string" && this.self.dataset.readOnly.toBoolean() !== readonly) {
+      if (readonly) {
+        this._lock();
```

Comment:
Here we had [this line](https://github.com/wbotelhos/raty/blob/master/lib/jquery.raty.js#L786):

```
this.self.off('.raty').children(this.opt.starType).off('.raty');
```

We didn't replace it, so I think the test will break.

Reply:
if we use pointer-event the no requirement to remove event listeners.

[MDN DOC](https://developer.mozilla.org/en-US/docs/Web/CSS/pointer-events)